internal/provider/resource_machine.go

```diff
@@ -80,12 +102,31 @@ func resourceMachineCreate(ctx context.Context, d *schema.ResourceData, meta int
 	constraints := d.Get("constraints").(string)
 	disks := d.Get("disks").(string)
 	series := d.Get("series").(string)
+	sshAddress := d.Get("ssh_address").(string)
+	publicKeyFile := d.Get("public_key_file").(string)
+	privateKeyFile := d.Get("private_key_file").(string)
+
+	if sshAddress != "" {
+		// Check and fail if any constraints are given along with manual
+		// provision directive
+		if constraints != "" {
+			return diag.FromErr(errors.NotValidf("Manual provision and constraints are mutually exclusive"))
+		}
+		// Check and fail if series is given along with manual provision
+		// directive
+		if series != "" {
+			return diag.FromErr(errors.NotValidf("Manual provision and constraints are mutually exclusive"))
```

Comment:
Change the message with "Manual provision and series are mutually exclusive"

Reply:
We should force private and public keys to be set.

Reply:
It would be much better if this is done declaratively in the schema. Something like:
```go
Schema: map[string]*schema.Schema{
			"name": {
				Description: "A name for the machine resource in Terraform.",
				Type:        schema.TypeString,
				Optional:    true,
				Computed:    true,
				ForceNew:    true,
			},
			"model": {
				Description: "The Juju model in which to add a new machine.",
				Type:        schema.TypeString,
				Required:    true,
				ForceNew:    true,
			},
			"constraints": {
				Description:   "Machine constraints that overwrite those available from 'juju get-model-constraints' and provider's defaults.",
				Type:          schema.TypeString,
				Optional:      true,
				Default:       "",
				ForceNew:      true,
				ConflictsWith: []string{"ssh_address"},
			},
			"disks": {
				Description:   "Storage constraints for disks to attach to the machine(s).",
				Type:          schema.TypeString,
				Optional:      true,
				Default:       "",
				ForceNew:      true,
				ConflictsWith: []string{"ssh_address"},
			},
			"series": {
				Description:   "The operating system series to install on the new machine(s).",
				Type:          schema.TypeString,
				Required:      false,
				Optional:      true,
				ForceNew:      true,
				ConflictsWith: []string{"ssh_address"},
			},
			"machine_id": {
				Description: "The id of the machine Juju creates.",
				Type:        schema.TypeString,
				Computed:    true,
				Optional:    false,
				Required:    false,
			},
			"ssh_address": {
				Description:  "The user@host directive for manual provisioning an existing machine via ssh.",
				Type:         schema.TypeString,
				Optional:     true,
				Default:      "",
				ForceNew:     true,
				RequiredWith: []string{"public_key_file", "private_key_file"},
			},
			"public_key_file": {
				Description: "The file path to read the public key from.",
				Type:        schema.TypeString,
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
			"private_key_file": {
				Description: "The file path to read the private key from.",
				Type:        schema.TypeString,
				Optional:    true,
				Default:     "",
				ForceNew:    true,
			},
		},
```